backend/src/server.rs

```diff
@@ -0,0 +1,501 @@
+            true
+        }
+    }).collect::<Vec<_>>();
+
+    if count == 0.0 {
+        return object! {
+            "first" => 0,
+            "last" => 0,
+            "min" => 0,
+            "max" => 0,
+            "mean" => 0,
+            "variance" => 0,
+            "trend" => 0,
+            "trend_b" => 0,
+            "n" => 0
+        };
+    }
+
+    let sums = data.iter().enumerate().map(|(i, d)| {
+        if skip_list[i] {
+            return 0.0;
+        }
+
+        let krate = &d.by_crate[crate_name];
+        let mut sum = 0.0;
+        for phase in phases.members() {
+            sum += krate[phase.as_str().unwrap()].time;
+        }
+        return sum;
+    }).collect::<Vec<_>>();
+
+    let first = sums[first];
+    let last = *sums.last().unwrap();
+
+    let mut min = first;
+    let mut max = first;
+    let q1_idx = (data.len() as f64) / 4.0;
+    let q4_idx = 3.0 * (data.len() as f64) / 4.0;
+    let mut total = 0.0;
+    let mut q1_total = 0.0;
+    let mut q4_total = 0.0;
+    for i in 0..data.len() {
+        if skip_list[i] {
+            continue;
+        }
+        let cur = sums[i];
+        total += cur;
+        min = min.min(cur);
+        max = max.max(cur);
+        if (i as f64) < q1_idx { // Within the first quartile
+            q1_total += cur;
+        }
+        if (i as f64) >= q4_idx { // Within the fourth quartile
+            q4_total += cur;
+        }
+    }
+
+    // Calculate the variance
+    let mean = total / count;
+    let mut var_total = 0.0;
+    for i in 0..data.len() {
+        if skip_list[i] {
+            continue;
+        }
+        let diff = sums[i] - mean;
+        var_total += diff * diff;
+    }
+    let variance = var_total / (count - 1.0);
+
+    let trend = if count >= 10.0 && count == (data.len() as f64) {
+        let q1_mean = q1_total / q1_idx;
+        let q4_mean = q4_total / ((data.len() as f64) - q4_idx);
+        100.0 * ((q4_mean - q1_mean) / first)
+    } else {
+        0.0
+    };
+    let trend_b = 100.0 * ((last - first) / first);
+
+    object! {
+        "first" => first,
+        "last" => last,
+        "min" => min,
+        "max" => max,
+        "mean" => mean,
+        "variance" => variance,
+        "trend" => trend,
+        "trend_b" => trend_b,
+        "n" => count
+    }
+}
+
+/// Pre and post processes the request and response to prepare it for the
+/// handler passed in. Specifically, it parses JSON data from the request if
+/// the request had greater than 0 length, and hands that into the handler.
+/// Post-processing consists of applying access control headers and [potentially]
+/// printing the error message.
+fn get_handler<F>(handler: F, req: &mut Request) -> IronResult<Response>
+    where F: Fn(Option<JsonValue>, &ServerData) -> Result<JsonValue> {
+    use std::ops::Deref;
+    use iron::headers::{ContentType, AccessControlAllowOrigin};
+    use iron::mime::{Mime, TopLevel, SubLevel};
+    use iron::modifiers::Header;
```

Comment:
> Prefer to put non-glob imports at the top level, rather than in functions
Do we want to follow that rule? All of these are very specific to this function (other than Deref).